maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/report/StatelessXmlReporter.java

```diff
@@ -234,71 +238,129 @@ private void serializeTestClassWithRerun( OutputStream outputStream, OutputStrea
+                                singleRunEntry.getReportEntryType().getRerunXmlTag(), true );
+                        }
                     }
-                }
-                break;
-            case error:
-            case failure:
-                // When rerunFailingTestsCount is set to larger than 0
-                startTestElement( ppw, firstMethodEntry );
-                boolean firstRun = true;
-                for ( WrappedReportEntry singleRunEntry : methodEntries )
-                {
-                    if ( firstRun )
+                    ppw.endElement();
+                    break;
+                case flake:
+                    startTestElement( ppw, firstMethodEntry );
+                    for ( WrappedReportEntry singleRunEntry : methodEntries )
                     {
-                        firstRun = false;
-                        getTestProblems( fw, ppw, singleRunEntry, trimStackTrace, outputStream,
+                        if ( firstRun )
+                        {
+                            firstRun = false;
+                            getTestProblems( fw, ppw, singleRunEntry, trimStackTrace, outputStream,
                                 singleRunEntry.getReportEntryType().getXmlTag(), false );
-                        createOutErrElements( fw, ppw, singleRunEntry, outputStream );
+                            createOutErrElements( fw, ppw, singleRunEntry, outputStream );
+                        }
+                        else
+                        {
+                            getTestProblems( fw, ppw, singleRunEntry, trimStackTrace, outputStream,
+                                singleRunEntry.getReportEntryType().getFlakyXmlTag(), true );
+                        }
+                    }
+                    ppw.endElement();
+                    break;
+                case skipped:
+                    startTestElement( ppw, firstMethodEntry );
+                    getTestProblems( fw, ppw, firstMethodEntry, trimStackTrace, outputStream,
+                        firstMethodEntry.getReportEntryType().getXmlTag(), false );
+                    ppw.endElement();
+                    break;
+                default:
+                    throw new IllegalStateException( "Get unknown test result type" );
+            }
+        }
+        else
+        {
+            switch ( getTestResultType( methodEntries ) )
+            {
+                case success:
+                    for ( WrappedReportEntry methodEntry : methodEntries )
+                    {
+                        if ( methodEntry.getReportEntryType() == SUCCESS )
+                        {
+                            startTestElement( ppw, methodEntry );
+                            ppw.endElement();
+                        }
                     }
-                    else
+                    break;
+                case error:
+                case failure:
```

Comment:
Does the code in this case repeat up above? If so, consider putting it into a helper method